provenance_lib/parse.py

```diff
@@ -0,0 +1,362 @@
+    # convention.
+
+    @property
+    def parents(self):
+        """ The list of ProvNodes used as inputs in creating this ProvNode """
+        # NOTE: We must delay gathering parentage data until the Archive is
+        # fully populated (or risk KeyError). Caching this lazily allows us to
+        # delay until the first call (likely the first DAG traversal)
+        if not self._parents:
+            try:
+                parent_dicts = [parent for parent in self._action.inputs]
+                parent_uuids = [
+                    list(uuid.values())[0] for uuid in parent_dicts]
+                self._parents = [
+                    self._origin_archives[0]._archive_contents[uuid]
+                    for uuid in parent_uuids]
+            except KeyError:
+                pass
+        return self._parents
+
+    @property
+    def uuid(self):
+        return self._result_md.uuid
+
+    @property
+    def sem_type(self):
+        return self._result_md.type
+
+    @property
+    def format(self):
+        return self._result_md.format
+
+    def __init__(self, origin_archive, zf: zipfile,
+                 fps_for_this_result: Iterator[pathlib.Path]):
+        self._origin_archives.append(origin_archive)
+        # TODO: Read and check VERSION
+        # (this will probably effect what other things get read in)
+        for fp in fps_for_this_result:
+            # TODO: Should we be reading these zipfiles once here,
+            # and then passing them to the constructors below?
+            if fp.name == 'metadata.yaml':
+                self._result_md = _ResultMetadata(zf, str(fp))
+            elif fp.name == 'action.yaml':
+                self._action = _Action(zf, str(fp))
+            elif fp.name == 'citations.bib':
+                self._citations = _Citations(zf, str(fp))
+
+    def __repr__(self):
+        return f'ProvNode({self.uuid}, {self.sem_type}, fmt={self.format})'
+
+    def __str__(self):
+        return f'ProvNode({self.uuid})'
+
+    def __eq__(self, other):
+        # TODO: Should this offer more robust validation?
+        return self.uuid == other.uuid
+
+    # TODO: Should this live in ProvDAG?
+    def traverse_uuids(self):
+        """ depth-first traversal of this ProvNode's ancestors """
+        local_parents = dict()
+        if not self.parents:
+            local_parents = {self.uuid: None}
+        else:
+            sub_dag = dict()
+            for parent in self.parents:
+                sub_dag.update(parent.traverse_uuids())
+            local_parents[self.uuid] = sub_dag
+        return local_parents
+
+
+class Archive:
+    """
+    Lightly-processed contents of a single QIIME 2 Archive, an Archive
+    contains a non-hierarchical pool of unique ProvNodes
+    """
+
+    # TODO: Can we assume all Q2 archives contain only unique Results?
+    # If not, Archive should filter duplicates.
+
+    # TODO: add property for archive version?
+    # TODO: UUID class with basic validation?
```

Comment:
Two possible paths here:

- Simple length and format check that this is a valid UUID
- More robust/expensive check of Archive validity and UUID correctness

I haven't looked into how feasible the second approach is, but if the UUID is a hash of the archive, we could re-hash from the passed archive and match the actual UUID string against the newly-generated hash, allowing us to check whether archives have been modified or corrupted. 

Recognizing that some users have valid reasons for manually hacking their Archives, this might be best as an optional feature.